yaml/_dumper_state.ts

```diff
@@ -781,69 +781,61 @@ export class DumperState {
       block = this.flowLevel < 0 || this.flowLevel > level;
     }
 
-    let duplicateIndex = -1;
-    let duplicate = false;
     if (isObject(value)) {
-      duplicateIndex = this.duplicates.indexOf(value);
-      duplicate = duplicateIndex !== -1;
-    }
-
-    if (
-      (tag !== null && tag !== "?") ||
-      duplicate ||
-      (this.indent !== 2 && level > 0)
-    ) {
-      compact = false;
-    }
+      const duplicateIndex = this.duplicates.indexOf(value);
+      const duplicate = duplicateIndex !== -1;
 
-    if (duplicate && this.usedDuplicates.has(value)) {
-      return `*ref_${duplicateIndex}`;
-    } else {
-      if (isObject(value)) {
-        if (duplicate) {
-          this.usedDuplicates.add(value);
-        }
-        if (!Array.isArray(value)) {
-          if (block && Object.keys(value).length !== 0) {
-            value = this.stringifyBlockMapping(value, { tag, level, compact });
-            if (duplicate) {
-              value = `&ref_${duplicateIndex}${value}`;
-            }
-          } else {
-            value = this.stringifyFlowMapping(value, { level });
-            if (duplicate) {
-              value = `&ref_${duplicateIndex} ${value}`;
-            }
+      if (duplicate) {
+        if (this.usedDuplicates.has(value)) return `*ref_${duplicateIndex}`;
+        this.usedDuplicates.add(value);
+      }
+      if (
+        (tag !== null && tag !== "?") ||
+        duplicate ||
+        (this.indent !== 2 && level > 0)
+      ) {
+        compact = false;
+      }
```

Comment:
Why was this moved?

Reply:
Because compact only needs to be evaluated after `if (this.usedDuplicates.has(value)) return ``*ref_${duplicateIndex}``;` doesn't return early.